File: labs OOP/1.2/C#/Trapezoid.cs
namespace TrapezoidLibCS
{
    public struct Point
    {
        public double X { get; }
        public double Y { get; }

        public Point(double x, double y)
        {
            X = x;
            Y = y;
        }
    }

    public class Trapezoid
    {
        private Point a, b, c, d;


        public Trapezoid()
        {
            a = new Point(0.0, 0.0);
            b = new Point(0.0, 0.0);
            c = new Point(0.0, 0.0);
            d = new Point(0.0, 0.0);
        }


        public Trapezoid(Point a, Point b, Point c, Point d)
        {
            this.a = a;
            this.b = b;
            this.c = c;
            this.d = d;
        }


        public Trapezoid(Trapezoid other) : this(other.a, other.b, other.c, other.d)
        {

        }


        ~Trapezoid()
        {

        }


        public double CalculateArea()
        {
            return 0.5 * System.Math.Abs((a.X * b.Y + b.X * c.Y + c.X * d.Y + d.X * a.Y) -
                                      (b.X * a.Y + c.X * b.Y + d.X * c.Y + a.X * d.Y));
        }


        public double CalculatePerimeter()
        {
            return SideLength(a, b) + SideLength(b, c) + SideLength(c, d) + SideLength(d, a);
        }

        // Геттери
        public Point GetA() { return a; }
        public Point GetB() { return b; }
        public Point GetC() { return c; }
        public Point GetD() { return d; }


        private double SideLength(Point p1, Point p2)
        {
            return System.Math.Sqrt(System.Math.Pow(p2.X - p1.X, 2) + System.Math.Pow(p2.Y - p1.Y, 2));
        }
    }
}

File: labs OOP/Lab 2.2/C++/Trapezoid.cpp
#include "Trapezoid.h"
#include <iostream>


Trapezoid::Trapezoid() : a{ 0.0, 0.0 }, b{ 0.0, 0.0 }, c{ 0.0, 0.0 }, d{ 0.0, 0.0 } {

}


Trapezoid::Trapezoid(Point a, Point b, Point c, Point d) : a(a), b(b), c(c), d(d) {

}


Trapezoid::Trapezoid(const Trapezoid& other) : a(other.a), b(other.b), c(other.c), d(other.d) {

}


Trapezoid::Trapezoid(Trapezoid&& other) noexcept : a(std::move(other.a)), b(std::move(other.b)), c(std::move(other.c)), d(std::move(other.d))
{

}


Trapezoid::~Trapezoid() {

}


double Trapezoid::calculateArea() const {
    return 0.5 * std::abs((a.x * b.y + b.x * c.y + c.x * d.y + d.x * a.y) -
        (b.x * a.y + c.x * b.y + d.x * c.y + a.x * d.y));
}


double Trapezoid::calculatePerimeter() const {
    return sideLength(a, b) + sideLength(b, c) + sideLength(c, d) + sideLength(d, a);
}


Point Trapezoid::getA() const { return a; }
Point Trapezoid::getB() const { return b; }
Point Trapezoid::getC() const { return c; }
Point Trapezoid::getD() const { return d; }


double Trapezoid::sideLength(Point p1, Point p2) const {
    return std::sqrt(std::pow(p2.x - p1.x, 2) + std::pow(p2.y - p1.y, 2));
}